In [ ]:
import gc
import pickle
import numpy as np 
import pandas as pd 
from category_encoders.ordinal import OrdinalEncoder
import matplotlib.pyplot as plt
from tsforest.utils import make_time_range
import holidays

import matplotlib.pyplot as plt
import seaborn as sns

# local modules
import sys
sys.path.append("../lib/")
from utils import compute_scaling, reduce_mem_usage

***
## data loading

In [ ]:
sales_train = pd.read_csv("../input/sales_train_validation.csv")
sales_train.info()

In [ ]:
calendar = pd.read_csv("../input/calendar.csv", parse_dates=["date"])
calendar.info()

In [ ]:
sell_prices = pd.read_csv("../input/sell_prices.csv")
sell_prices.info()

***
### Events features

In [ ]:
us_holidays = holidays.UnitedStates(years=[2011,2012,2013,2014,2015,2016], observed=True)
holidays_dataframe = pd.DataFrame(us_holidays.items(), columns=["ds","event"])
holidays_dataframe.sort_values("ds", inplace=True)

***
### long weekends features

In [5]:
lw_features = make_time_range("2011-01-29", "2016-06-19", "D")
lw_features.set_index("ds", inplace=True)
lw_features["lw_type"] = 0
lw_features["lw_day"] = 0
lw_features["prev_lw"] = 0
lw_features["post_lw"] = 0

In [6]:
# 2011-02-19 al 2011-02-21
lw_features.loc['2011-02-18', 'prev_lw'] = 1
lw_features.loc['2011-02-19':'2011-02-21', 'lw_type'] = 1
lw_features.loc['2011-02-19':'2011-02-21', 'lw_day'] = [1,2,3]
lw_features.loc['2011-02-22', 'post_lw'] = 1

# 2011-05-28 al 2011-05-30
lw_features.loc['2011-05-27', 'prev_lw'] = 1
lw_features.loc['2011-05-28':'2011-05-30', 'lw_type'] = 1
lw_features.loc['2011-05-28':'2011-05-30', 'lw_day'] = [1,2,3]
lw_features.loc['2011-05-31', 'post_lw'] = 1

# 2011-07-02 al 2011-07-04
lw_features.loc['2011-07-01', 'prev_lw'] = 1
lw_features.loc['2011-07-02':'2011-07-04', 'lw_type'] = 1
lw_features.loc['2011-07-02':'2011-07-04', 'lw_day'] = [1,2,3]
lw_features.loc['2011-07-05', 'post_lw'] = 1

# 2011-09-03 al 2011-09-05
lw_features.loc['2011-09-02', 'prev_lw'] = 1
lw_features.loc['2011-09-03':'2011-09-05', 'lw_type'] = 1
lw_features.loc['2011-09-03':'2011-09-05', 'lw_day'] = [1,2,3]
lw_features.loc['2011-09-06', 'post_lw'] = 1

# 2011-10-08 al 2011-10-10
lw_features.loc['2011-10-07', 'prev_lw'] = 1
lw_features.loc['2011-10-08':'2011-10-10', 'lw_type'] = 1
lw_features.loc['2011-10-08':'2011-10-10', 'lw_day'] = [1,2,3]
lw_features.loc['2011-10-11', 'post_lw'] = 1

# 2011-11-11 al 2011-11-13
lw_features.loc['2011-11-10', 'prev_lw'] = 1
lw_features.loc['2011-11-11':'2011-11-13', 'lw_type'] = 2
lw_features.loc['2011-11-11':'2011-11-13', 'lw_day'] = [1,2,3]
lw_features.loc['2011-11-14', 'post_lw'] = 1

# 2011-12-24 al 2011-12-26
lw_features.loc['2011-12-23', 'prev_lw'] = 1
lw_features.loc['2011-12-24':'2011-12-26', 'lw_type'] = 1
lw_features.loc['2011-12-24':'2011-12-26', 'lw_day'] = [1,2,3]
lw_features.loc['2011-12-27', 'post_lw'] = 1

# 2011-12-31 al 2012-01-02
lw_features.loc['2011-12-30', 'prev_lw'] = 1
lw_features.loc['2011-12-31':'2012-01-02', 'lw_type'] = 1
lw_features.loc['2011-12-31':'2012-01-02', 'lw_day'] = [1,2,3]
lw_features.loc['2012-01-03', 'post_lw'] = 1

# 2012-01-14 al 2012-01-16
lw_features.loc['2012-01-13', 'prev_lw'] = 1
lw_features.loc['2012-01-14':'2012-01-16', 'lw_type'] = 1
lw_features.loc['2012-01-14':'2012-01-16', 'lw_day'] = [1,2,3]
lw_features.loc['2012-01-17', 'post_lw'] = 1

# 2012-02-18 al 2012-02-20
lw_features.loc['2012-02-17', 'prev_lw'] = 1
lw_features.loc['2012-02-18':'2012-02-20', 'lw_type'] = 1
lw_features.loc['2012-02-18':'2012-02-20', 'lw_day'] = [1,2,3]
lw_features.loc['2012-02-21', 'post_lw'] = 1

# 2012-05-26 al 2012-05-28
lw_features.loc['2012-05-25', 'prev_lw'] = 1
lw_features.loc['2012-05-26':'2012-05-28', 'lw_type'] = 1
lw_features.loc['2012-05-26':'2012-05-28', 'lw_day'] = [1,2,3]
lw_features.loc['2012-05-29', 'post_lw'] = 1

# 2012-09-01 al 2012-09-03
lw_features.loc['2012-08-31', 'prev_lw'] = 1
lw_features.loc['2012-09-01':'2012-09-03', 'lw_type'] = 1
lw_features.loc['2012-09-01':'2012-09-03', 'lw_day'] = [1,2,3]
lw_features.loc['2012-09-04', 'post_lw'] = 1

# 2012-10-06 al 2012-10-08
lw_features.loc['2012-10-05', 'prev_lw'] = 1
lw_features.loc['2012-10-06':'2012-10-08', 'lw_type'] = 1
lw_features.loc['2012-10-06':'2012-10-08', 'lw_day'] = [1,2,3]
lw_features.loc['2012-10-09', 'post_lw'] = 1

# 2013-01-19 al 2013-01-21
lw_features.loc['2013-01-18', 'prev_lw'] = 1
lw_features.loc['2013-01-19':'2013-01-21', 'lw_type'] = 1
lw_features.loc['2013-01-19':'2013-01-21', 'lw_day'] = [1,2,3]
lw_features.loc['2013-01-22', 'post_lw'] = 1

# 2013-02-16 al 2013-02-18
lw_features.loc['2013-02-15', 'prev_lw'] = 1
lw_features.loc['2013-02-16':'2013-02-18', 'lw_type'] = 1
lw_features.loc['2013-02-16':'2013-02-18', 'lw_day'] = [1,2,3]
lw_features.loc['2013-02-19', 'post_lw'] = 1

# 2013-05-25 al 2013-05-27
lw_features.loc['2013-05-24', 'prev_lw'] = 1
lw_features.loc['2013-05-25':'2013-05-27', 'lw_type'] = 1
lw_features.loc['2013-05-25':'2013-05-27', 'lw_day'] = [1,2,3]
lw_features.loc['2013-05-28', 'post_lw'] = 1

# 2013-08-31 al 2013-09-02
lw_features.loc['2013-08-30', 'prev_lw'] = 1
lw_features.loc['2013-08-31':'2013-09-02', 'lw_type'] = 1
lw_features.loc['2013-08-31':'2013-09-02', 'lw_day'] = [1,2,3]
lw_features.loc['2013-09-03', 'post_lw'] = 1

# 2013-10-12 al 2013-10-14
lw_features.loc['2013-10-11', 'prev_lw'] = 1
lw_features.loc['2013-10-12':'2013-10-14', 'lw_type'] = 1
lw_features.loc['2013-10-12':'2013-10-14', 'lw_day'] = [1,2,3]
lw_features.loc['2013-10-15', 'post_lw'] = 1

# 2013-11-09 al 2013-11-11
lw_features.loc['2013-11-08', 'prev_lw'] = 1
lw_features.loc['2013-11-09':'2013-11-11', 'lw_type'] = 1
lw_features.loc['2013-11-09':'2013-11-11', 'lw_day'] = [1,2,3]
lw_features.loc['2013-11-12', 'post_lw'] = 1

# 2014-01-18 al 2014-01-20
lw_features.loc['2014-01-17', 'prev_lw'] = 1
lw_features.loc['2014-01-18':'2014-01-20', 'lw_type'] = 1
lw_features.loc['2014-01-18':'2014-01-20', 'lw_day'] = [1,2,3]
lw_features.loc['2014-01-21', 'post_lw'] = 1

# 2014-02-15 al 2014-02-17
lw_features.loc['2014-02-14', 'prev_lw'] = 1
lw_features.loc['2014-02-15':'2014-02-17', 'lw_type'] = 1
lw_features.loc['2014-02-15':'2014-02-17', 'lw_day'] = [1,2,3]
lw_features.loc['2014-02-18', 'post_lw'] = 1

# 2014-05-24 al 2014-05-26
lw_features.loc['2014-05-23', 'prev_lw'] = 1
lw_features.loc['2014-05-24':'2014-05-26', 'lw_type'] = 1
lw_features.loc['2014-05-24':'2014-05-26', 'lw_day'] = [1,2,3]
lw_features.loc['2014-05-27', 'post_lw'] = 1

# 2014-07-04 al 2014-07-06
lw_features.loc['2014-07-03', 'prev_lw'] = 1
lw_features.loc['2014-07-04':'2014-07-06', 'lw_type'] = 2
lw_features.loc['2014-07-04':'2014-07-06', 'lw_day'] = [1,2,3]
lw_features.loc['2014-07-07', 'post_lw'] = 1

# 2014-08-30 al 2014-09-01
lw_features.loc['2014-08-29', 'prev_lw'] = 1
lw_features.loc['2014-08-30':'2014-09-01', 'lw_type'] = 1
lw_features.loc['2014-08-30':'2014-09-01', 'lw_day'] = [1,2,3]
lw_features.loc['2014-09-02', 'post_lw'] = 1

# 2014-10-11 al 2014-10-13
lw_features.loc['2014-10-10', 'prev_lw'] = 1
lw_features.loc['2014-10-11':'2014-10-13', 'lw_type'] = 1
lw_features.loc['2014-10-11':'2014-10-13', 'lw_day'] = [1,2,3]
lw_features.loc['2014-10-14', 'post_lw'] = 1

# 2015-01-17 al 2015-01-19
lw_features.loc['2015-01-16', 'prev_lw'] = 1
lw_features.loc['2015-01-17':'2015-01-19', 'lw_type'] = 1
lw_features.loc['2015-01-17':'2015-01-19', 'lw_day'] = [1,2,3]
lw_features.loc['2015-01-20', 'post_lw'] = 1

# 2015-02-14 al 2015-02-16
lw_features.loc['2015-02-13', 'prev_lw'] = 1
lw_features.loc['2015-02-14':'2015-02-16', 'lw_type'] = 1
lw_features.loc['2015-02-14':'2015-02-16', 'lw_day'] = [1,2,3]
lw_features.loc['2015-02-17', 'post_lw'] = 1

# 2015-05-23 al 2015-05-25
lw_features.loc['2015-05-22', 'prev_lw'] = 1
lw_features.loc['2015-05-23':'2015-05-25', 'lw_type'] = 1
lw_features.loc['2015-05-23':'2015-05-25', 'lw_day'] = [1,2,3]
lw_features.loc['2015-05-26', 'post_lw'] = 1

# 2015-07-03 al 2015-07-05
lw_features.loc['2015-07-02', 'prev_lw'] = 1
lw_features.loc['2015-07-03':'2015-07-05', 'lw_type'] = 2
lw_features.loc['2015-07-03':'2015-07-05', 'lw_day'] = [1,2,3]
lw_features.loc['2015-07-06', 'post_lw'] = 1

# 2015-09-05 al 2015-09-07
lw_features.loc['2015-09-04', 'prev_lw'] = 1
lw_features.loc['2015-09-05':'2015-09-07', 'lw_type'] = 1
lw_features.loc['2015-09-05':'2015-09-07', 'lw_day'] = [1,2,3]
lw_features.loc['2015-09-08', 'post_lw'] = 1

# 2015-10-10 al 2015-10-12
lw_features.loc['2015-10-09', 'prev_lw'] = 1
lw_features.loc['2015-10-10':'2015-10-12', 'lw_type'] = 1
lw_features.loc['2015-10-10':'2015-10-12', 'lw_day'] = [1,2,3]
lw_features.loc['2015-10-13', 'post_lw'] = 1

# 2015-12-25 al 2015-12-27
lw_features.loc['2015-12-24', 'prev_lw'] = 1
lw_features.loc['2015-12-25':'2015-12-27', 'lw_type'] = 2
lw_features.loc['2015-12-25':'2015-12-27', 'lw_day'] = [1,2,3]
lw_features.loc['2015-12-28', 'post_lw'] = 1

# 2016-01-01 al 2016-01-03
lw_features.loc['2015-12-31', 'prev_lw'] = 1
lw_features.loc['2016-01-01':'2016-01-03', 'lw_type'] = 2
lw_features.loc['2016-01-01':'2016-01-03', 'lw_day'] = [1,2,3]
lw_features.loc['2016-01-04', 'post_lw'] = 1

# 2016-01-16 al 2016-01-18
lw_features.loc['2016-01-15', 'prev_lw'] = 1
lw_features.loc['2016-01-16':'2016-01-18', 'lw_type'] = 1
lw_features.loc['2016-01-16':'2016-01-18', 'lw_day'] = [1,2,3]
lw_features.loc['2016-01-19', 'post_lw'] = 1

# 2016-02-13 al 2016-02-15
lw_features.loc['2016-02-12', 'prev_lw'] = 1
lw_features.loc['2016-02-13':'2016-02-15', 'lw_type'] = 1
lw_features.loc['2016-02-13':'2016-02-15', 'lw_day'] = [1,2,3]
lw_features.loc['2016-02-16', 'post_lw'] = 1

# 2016-05-28 al 2016-05-30
lw_features.loc['2016-05-27', 'prev_lw'] = 1
lw_features.loc['2016-05-28':'2016-05-30', 'lw_type'] = 1
lw_features.loc['2016-05-28':'2016-05-30', 'lw_day'] = [1,2,3]
lw_features.loc['2016-05-31', 'post_lw'] = 1


In [7]:
lw_features.reset_index(inplace=True)

In [8]:
lw_features.head()

ds  lw_type  lw_day  prev_lw  post_lw
0 2011-01-29        0       0        0        0
1 2011-01-30        0       0        0        0
2 2011-01-31        0       0        0        0
3 2011-02-01        0       0        0        0
4 2011-02-02        0       0        0        0

***
### events features

In [9]:
events_features = make_time_range("2011-01-29", "2016-06-19", "D")

In [10]:
# christmas features
events_features["prev_christmas"] = 0
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 20").index, "prev_christmas"] = 1
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 21").index, "prev_christmas"] = 2
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 22").index, "prev_christmas"] = 3
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 23").index, "prev_christmas"] = 4
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 24").index, "prev_christmas"] = 5
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 25").index, "prev_christmas"] = 6

events_features["post_christmas"] = 0
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 26").index, "post_christmas"] = 1

In [11]:
# new year features
events_features["prev_newyear"] = 0
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 28").index, "prev_newyear"] = 1
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 29").index, "prev_newyear"] = 2
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 30").index, "prev_newyear"] = 3
events_features.loc[events_features.query("ds.dt.month == 12 & ds.dt.day == 31").index, "prev_newyear"] = 4

events_features["post_newyear"] = 0
events_features.loc[events_features.query("ds.dt.month == 1 & ds.dt.day == 1").index, "post_newyear"] = 1

In [12]:
# Thanksgiving features
events_features["prev_thanksgiving"] = 0
events_features["post_thanksgiving"] = 0

idx = events_features.query("ds.dt.year == 2011 & ds.dt.month == 11 & ds.dt.day in [22,23,24]").index
events_features.loc[idx, "prev_thanksgiving"] = [1,2,3]
idx = events_features.query("ds.dt.year == 2011 & ds.dt.month == 11 & ds.dt.day == 25").index
events_features.loc[idx, "post_thanksgiving"] = 1

idx = events_features.query("ds.dt.year == 2012 & ds.dt.month == 11 & ds.dt.day in [20,21,22]").index
events_features.loc[idx, "prev_thanksgiving"] = [1,2,3]
idx = events_features.query("ds.dt.year == 2012 & ds.dt.month == 11 & ds.dt.day == 23").index
events_features.loc[idx, "post_thanksgiving"] = 1

idx = events_features.query("ds.dt.year == 2013 & ds.dt.month == 11 & ds.dt.day in [26,27,28]").index
events_features.loc[idx, "prev_thanksgiving"] = [1,2,3]
idx = events_features.query("ds.dt.year == 2013 & ds.dt.month == 11 & ds.dt.day == 29").index
events_features.loc[idx, "post_thanksgiving"] = 1

idx = events_features.query("ds.dt.year == 2014 & ds.dt.month == 11 & ds.dt.day in [25,26,27]").index
events_features.loc[idx, "prev_thanksgiving"] = [1,2,3]
idx = events_features.query("ds.dt.year == 2014 & ds.dt.month == 11 & ds.dt.day == 28").index
events_features.loc[idx, "post_thanksgiving"] = 1

idx = events_features.query("ds.dt.year == 2015 & ds.dt.month == 11 & ds.dt.day in [24,25,26]").index
events_features.loc[idx, "prev_thanksgiving"] = [1,2,3]
idx = events_features.query("ds.dt.year == 2015 & ds.dt.month == 11 & ds.dt.day == 27").index
events_features.loc[idx, "post_thanksgiving"] = 1

In [13]:
events_features.head()

ds  prev_christmas  post_christmas  prev_newyear  post_newyear  \
0 2011-01-29               0               0             0             0   
1 2011-01-30               0               0             0             0   
2 2011-01-31               0               0             0             0   
3 2011-02-01               0               0             0             0   
4 2011-02-02               0               0             0             0   

   prev_thanksgiving  post_thanksgiving  
0                  0                  0  
1                  0                  0  
2                  0                  0  
3                  0                  0  
4                  0                  0

***
## hierarchy

In [14]:
sales_train["id"] = sales_train.id.map(lambda x: x.replace("_validation", ""))
hierarchy = (sales_train.loc[:, ["id", "item_id", "dept_id", "cat_id", "store_id", "state_id"]]
             .drop_duplicates())
encoders = dict()

In [15]:
hierarchy.to_parquet("../input/hierarchy_raw.parquet", index=False)

In [16]:
# hierarchy encoder
id_encoder = OrdinalEncoder()
id_encoder.fit(hierarchy.loc[:, ["id"]])
hierarchy["ts_id"]  = id_encoder.transform(hierarchy.loc[:, ["id"]])
encoders["id"] = id_encoder

item_encoder = OrdinalEncoder()
item_encoder.fit(hierarchy.loc[:, ["item_id"]])
hierarchy.loc[:, "item_id"]  = item_encoder.transform(hierarchy.loc[:, ["item_id"]])
encoders["item"] = item_encoder

dept_encoder = OrdinalEncoder()
dept_encoder.fit(hierarchy.loc[:, ["dept_id"]])
hierarchy.loc[:, "dept_id"]  = dept_encoder.transform(hierarchy.loc[:, ["dept_id"]])
encoders["dept"] = dept_encoder

cat_encoder = OrdinalEncoder()
cat_encoder.fit(hierarchy.loc[:, ["cat_id"]])
hierarchy.loc[:, "cat_id"]   = cat_encoder.transform(hierarchy.loc[:, ["cat_id"]])
encoders["cat"] = cat_encoder

store_encoder = OrdinalEncoder()
store_encoder.fit(hierarchy.loc[:, ["store_id"]])
hierarchy.loc[:, "store_id"] = store_encoder.transform(hierarchy.loc[:, ["store_id"]])
encoders["store"] = store_encoder

state_encoder = OrdinalEncoder()
state_encoder.fit(hierarchy.loc[:, ["state_id"]])
hierarchy.loc[:, "state_id"] = state_encoder.transform(hierarchy.loc[:, ["state_id"]])
encoders["state"] = state_encoder

In [17]:
hierarchy.to_parquet("../input/hierarchy.parquet", index=False)

In [18]:
outfile = open("../input/encoders.pkl", "wb")
pickle.dump(encoders, outfile)
outfile.close()

***
## calendar events encoding

In [19]:
event_name_1_encoder = OrdinalEncoder()
event_name_1_encoder.fit(calendar.loc[:, ["event_name_1"]])
calendar.loc[:, "event_name_1"] = event_name_1_encoder.transform(calendar.loc[:, ["event_name_1"]])

In [20]:
event_type_1_encoder = OrdinalEncoder()
event_type_1_encoder.fit(calendar.loc[:, ["event_type_1"]])
calendar.loc[:, "event_type_1"] = event_type_1_encoder.transform(calendar.loc[:, ["event_type_1"]])

In [21]:
event_name_2_encoder = OrdinalEncoder()
event_name_2_encoder.fit(calendar.loc[:, ["event_name_2"]])
calendar.loc[:, "event_name_2"] = event_name_2_encoder.transform(calendar.loc[:, ["event_name_2"]])

In [22]:
event_type_2_encoder = OrdinalEncoder()
event_type_2_encoder.fit(calendar.loc[:, ["event_type_2"]])
calendar.loc[:, "event_type_2"] = event_type_2_encoder.transform(calendar.loc[:, ["event_type_2"]])

***
## categorical encoding

In [23]:
sales_train["ts_id"] = id_encoder.transform(sales_train.loc[:, ["id"]])
sales_train.loc[:, "item_id"]  = item_encoder.transform(sales_train.loc[:, ["item_id"]])
sales_train.loc[:, "dept_id"]  = dept_encoder.transform(sales_train.loc[:, ["dept_id"]])
sales_train.loc[:, "cat_id"]   = cat_encoder.transform(sales_train.loc[:, ["cat_id"]])
sales_train.loc[:, "store_id"] = store_encoder.transform(sales_train.loc[:, ["store_id"]])
sales_train.loc[:, "state_id"] = state_encoder.transform(sales_train.loc[:, ["state_id"]])

In [24]:
sell_prices.loc[:, "store_id"] = store_encoder.transform(sell_prices.loc[:, ["store_id"]])
sell_prices.loc[:, "item_id"]  = item_encoder.transform(sell_prices.loc[:, ["item_id"]]) 

***
## building price features

In [25]:
number_prices = (sell_prices
                 .groupby(["store_id", "item_id"])["sell_price"]
                 .apply(lambda x: len(np.unique(x)))
                 .reset_index(name="n_prices")
                )

In [26]:
number_prices.n_prices.describe()

count    30490.000000
mean         2.769367
std          1.893627
min          1.000000
25%          1.000000
50%          2.000000
75%          4.000000
max         21.000000
Name: n_prices, dtype: float64

In [27]:
sell_prices.wm_yr_wk.nunique()

282

La mayoría de item-stores no tiene más de 4 precios diferentes, y el 50% no tiene más de 2 precios. El que más tiene, tiene 21 precios sobre un rango de 282 semanas.

In [28]:
sell_prices.query("item_id == 1 & store_id == 1").sell_price.value_counts()

8.26    141
8.38     11
9.58      2
Name: sell_price, dtype: int64

In [29]:
# identifies the more frequent price
regular_prices = (
    sell_prices
    .groupby(["store_id", "item_id"])["sell_price"]
    .apply(lambda x: x.value_counts().index[0])
    .reset_index(name="regular_price")
)

In [30]:
# price variation
price_variation1 = (sell_prices.groupby(["store_id","item_id"])["sell_price"].quantile(0.75) - 
                    sell_prices.groupby(["store_id","item_id"])["sell_price"].quantile(0.25)).reset_index(name="price_iqr1")

price_variation2 = (sell_prices.groupby(["store_id","item_id"])["sell_price"].quantile(0.95) - 
                    sell_prices.groupby(["store_id","item_id"])["sell_price"].quantile(0.05)).reset_index(name="price_iqr2")

# min and max prices
price_min = sell_prices.groupby(["store_id","item_id"])["sell_price"].quantile(0.025).reset_index(name="price_min")
price_max = sell_prices.groupby(["store_id","item_id"])["sell_price"].quantile(0.975).reset_index(name="price_max")

In [31]:
sell_prices = (
    sell_prices
    .merge(number_prices, how="left")
    .merge(regular_prices, how="left")
    .merge(price_variation1, how="left")
    .merge(price_variation2, how="left")
    .merge(price_min, how="left")
    .merge(price_max, how="left")
    .assign(discount = lambda x: x.regular_price - x.sell_price)
    .assign(discount_norm = lambda x: (x.regular_price - x.sell_price)/x.regular_price)
)

In [32]:
# price momentum
sell_prices = sell_prices.merge(calendar.loc[:, ['wm_yr_wk','month','year']], on=['wm_yr_wk'], how='left')
sell_prices['price_momentum_m'] = sell_prices['sell_price']/sell_prices.groupby(['store_id','item_id','month'])['sell_price'].transform('mean')
sell_prices['price_momentum_y'] = sell_prices['sell_price']/sell_prices.groupby(['store_id','item_id','year'])['sell_price'].transform('mean')
sell_prices.drop(["year","month"], axis=1, inplace=True)

In [33]:
sell_prices

store_id  item_id  wm_yr_wk  sell_price  n_prices  regular_price  \
0                1        1     11325        9.58         3           8.26   
1                1        1     11325        9.58         3           8.26   
2                1        1     11325        9.58         3           8.26   
3                1        1     11325        9.58         3           8.26   
4                1        1     11325        9.58         3           8.26   
...            ...      ...       ...         ...       ...            ...   
47735392        10     3049     11620        1.00         1           1.00   
47735393        10     3049     11620        1.00         1           1.00   
47735394        10     3049     11620        1.00         1           1.00   
47735395        10     3049     11621        1.00         1           1.00   
47735396        10     3049     11621        1.00         1           1.00   

          price_iqr1  price_iqr2  price_min  price_max  discount  \
0                0.0        0.12       8.26       8.38     -1.32   
1                0.0        0.12       8.26       8.38     -1.32   
2                0.0        0.12       8.26       8.38     -1.32   
3                0.0        0.12       8.26       8.38     -1.32   
4                0.0        0.12       8.26       8.38     -1.32   
...              ...         ...        ...        ...       ...   
47735392         0.0        0.00       1.00       1.00      0.00   
47735393         0.0        0.00       1.00       1.00      0.00   
47735394         0.0        0.00       1.00       1.00      0.00   
47735395         0.0        0.00       1.00       1.00      0.00   
47735396         0.0        0.00       1.00       1.00      0.00   

          discount_norm  price_momentum_m  price_momentum_y  
0             -0.159806          1.128633          1.144914  
1             -0.159806          1.128633          1.144914  
2             -0.159806          1.128633          1.144914  
3             -0.159806          1.128633          1.144914  
4             -0.159806          1.128633          1.144914  
...                 ...               ...               ...  
47735392       0.000000          1.000000          1.000000  
47735393       0.000000          1.000000          1.000000  
47735394       0.000000          1.000000          1.000000  
47735395       0.000000          1.000000          1.000000  
47735396       0.000000          1.000000          1.000000  

[47735397 rows x 14 columns]

***
## data wrangling

In [34]:
data = pd.melt(sales_train, 
               id_vars=["ts_id","item_id","dept_id","cat_id","store_id","state_id"],
               value_vars=[f"d_{i}" for i in range(1,1914)],
               var_name="d",
               value_name="q")

In [35]:
calendar_columns = ["date", "wm_yr_wk", "d", "snap_CA", "snap_TX", "snap_WI",
                    "event_name_1", "event_type_1", "event_name_2", "event_type_2"]
data = pd.merge(data, 
                calendar.loc[:, calendar_columns],
                how="left",
                on="d")

In [36]:
data = pd.merge(data, sell_prices,
                on=["store_id", "item_id", "wm_yr_wk"],
                how="left")

MemoryError: Unable to allocate 22.4 GiB for an array with shape (9, 334495112) and data type float64

In [ ]:
data.rename({"date":"ds"}, axis=1, inplace=True)

In [ ]:
data = (data
        .merge(events_features, how="left", on=["ds"])
        .merge(lw_features, how="left", on=["ds"]))

In [ ]:
data.sort_values(["item_id","store_id","ds"], inplace=True, ignore_index=True)

***
## reduction of span features

In [ ]:
data["snap"] = 0

idx_snap_ca = data.query("state_id==1 & snap_CA==1").index
data.loc[idx_snap_ca, "snap"] = 1

idx_snap_tx = data.query("state_id==2 & snap_TX==1").index
data.loc[idx_snap_tx, "snap"] = 2

idx_snap_wi = data.query("state_id==3 & snap_WI==1").index
data.loc[idx_snap_wi, "snap"] = 3

In [ ]:
data.drop(["snap_CA", "snap_TX", "snap_WI"], axis=1, inplace=True)

***
## cleaning


### removes zeros at the start of the time series

In [ ]:
plt.figure(figsize=(15,5))
sns.distplot(data.query("q <= 20").q.values, kde=False)
plt.grid()
plt.show()

In [ ]:
def remove_starting_zeros(dataframe):
    idxmin = dataframe.query("q > 0").index.min()
    return dataframe.loc[idxmin:, :]

In [ ]:
%%time
data = (data
        .groupby(["item_id","store_id"])
        .apply(remove_starting_zeros)
        .reset_index(drop=True))

In [ ]:
plt.figure(figsize=(15,5))
sns.distplot(data.query("q <= 20").q.values, kde=False)
plt.grid()
plt.show()

***
## out of stock patterns

In [ ]:
def find_out_of_stock(df, threshold=28):
    df = df.copy()
    df["no_stock"] = 0
    zero_mask = (df.q == 0)
    transition_mask = (zero_mask != zero_mask.shift(1))
    zero_sequences = transition_mask.cumsum()[zero_mask]
    idx = zero_sequences[zero_sequences.map(zero_sequences.value_counts()) >= threshold].index 
    df.loc[idx, "no_stock"] = 1
    return df

In [ ]:
%%time
data = data.groupby(["item_id","store_id"]).apply(find_out_of_stock, threshold=28)
data.reset_index(drop=True, inplace=True)

***

In [ ]:
data = reduce_mem_usage(data)
gc.collect()

In [ ]:
data.drop(["d", "wm_yr_wk"], axis=1, inplace=True)

***

In [ ]:
# if the ts has no sales in the past 8 weeks, will be marked as out-of-stock for the prediction period.
left_date = data.ds.max() - pd.DateOffset(days=56)
no_sales_ts = (data
               .query("ds >= @left_date")
               .groupby(["ts_id"])
               .filter(lambda x: np.all(x.q==0))
               .loc[:, ["ts_id"]]
               .drop_duplicates()
              )

***
## validation and evaluation dataframes

In [ ]:
calendar_columns = ["date", "wm_yr_wk", "snap_CA", "snap_TX", "snap_WI",
                    "event_name_1", "event_type_1", "event_name_2", "event_type_2"]

valid_dataframe = (pd.concat([make_time_range("2016-04-25", "2016-05-22", "D").assign(**row)
                              for _,row in hierarchy.iterrows()], ignore_index=True)
                   .merge(calendar.loc[:, calendar_columns],
                          how="left", left_on="ds", right_on="date")
                   .merge(sell_prices, how="left")
                   .merge(events_features, how="left", on="ds")
                   .merge(lw_features, how="left", on="ds")
                   .drop(["id","date","wm_yr_wk"], axis=1)
                  )

In [ ]:
valid_dataframe["snap"] = 0

idx_snap_ca = valid_dataframe.query("state_id==1 & snap_CA==1").index
valid_dataframe.loc[idx_snap_ca, "snap"] = 1

idx_snap_tx = valid_dataframe.query("state_id==2 & snap_TX==1").index
valid_dataframe.loc[idx_snap_tx, "snap"] = 2

idx_snap_wi = valid_dataframe.query("state_id==3 & snap_WI==1").index
valid_dataframe.loc[idx_snap_wi, "snap"] = 3

valid_dataframe.drop(["snap_CA", "snap_TX", "snap_WI"], axis=1, inplace=True)

In [ ]:
valid_dataframe["no_stock"] = 0
idx = valid_dataframe.query("ts_id in @no_sales_ts.ts_id").index
valid_dataframe.loc[idx, "no_stock"] = 1

In [ ]:
valid_dataframe = reduce_mem_usage(valid_dataframe)

***

In [ ]:
calendar_columns = ["date", "wm_yr_wk", "snap_CA", "snap_TX", "snap_WI",
                    "event_name_1", "event_type_1", "event_name_2", "event_type_2"]

eval_dataframe = (pd.concat([make_time_range("2016-05-23", "2016-06-19", "D").assign(**row)
                             for _,row in hierarchy.iterrows()], ignore_index=True)
                  .merge(calendar.loc[:, calendar_columns],
                         how="left", left_on="ds", right_on="date")
                  .merge(sell_prices, how="left")
                  .merge(events_features, how="left", on="ds")
                  .merge(lw_features, how="left", on="ds")
                  .drop(["id","date","wm_yr_wk"], axis=1)
                 )

In [ ]:
eval_dataframe["snap"] = 0

idx_snap_ca = eval_dataframe.query("state_id==1 & snap_CA==1").index
eval_dataframe.loc[idx_snap_ca, "snap"] = 1

idx_snap_tx = eval_dataframe.query("state_id==2 & snap_TX==1").index
eval_dataframe.loc[idx_snap_tx, "snap"] = 2

idx_snap_wi = eval_dataframe.query("state_id==3 & snap_WI==1").index
eval_dataframe.loc[idx_snap_wi, "snap"] = 3

eval_dataframe.drop(["snap_CA", "snap_TX", "snap_WI"], axis=1, inplace=True)

In [ ]:
eval_dataframe["no_stock"] = 0
idx = eval_dataframe.query("ts_id in @no_sales_ts.ts_id").index
eval_dataframe.loc[idx, "no_stock"] = 1

In [ ]:
eval_dataframe = reduce_mem_usage(eval_dataframe)

***
### Saving the dataframes

In [ ]:
# training data
(data
 .to_parquet("../input/train_dataframe.parquet", index=False)
)

In [ ]:
# validation data
(valid_dataframe
 .to_parquet("../input/valid_dataframe.parquet", index=False)
)

In [ ]:
# evaluation data
(eval_dataframe
 .to_parquet("../input/eval_dataframe.parquet", index=False)
)

***